Repeated regex strings cause exception if "Only use first match" is not checked

We have a build that runs a python script to execute parameterized tests. These tests output many of the same lines for each run of the test. i.e:
```
Creating files for run 0
Creating files for run 100
Creating files for run 200
.... [more output] ...
Creating files for run 0
Creating files for run 100
Creating files for run 200
.... [more output] ...
```
My expectation was to to parse the time duration for each of these parameterized tests by using the regex `.*Creating files for run 0.*` , label "Run Test" (following the `.*Started by.*` example). However, if the "Only use first match" is *not* checked for that regex, I get an exception (shown below). If "Only use first match is checked, there is no exception. I suppose I was expecting a "Run Test 0", "Run Test 1", ... "Run Test N" where N is the number of times that expression was seen to be parsed and placed on the graph. If this is not the intended use, please let me know, though it certainly seems useful. 

The Exception referenced above: 
```
javax.servlet.ServletException: org.apache.commons.jelly.JellyTagException: jar:file:/var/lib/jenkins/plugins/build-time-blame/WEB-INF/lib/build-time-blame.jar!/org/jenkins/ci/plugins/buildtimeblame/action/BlameAction/index.jelly:11:70: <st:include> No page found 'body.jelly' for class org.jenkins.ci.plugins.buildtimeblame.action.BlameAction
	at org.kohsuke.stapler.jelly.JellyClassTearOff.serveIndexJelly(JellyClassTearOff.java:117)
	at org.kohsuke.stapler.jelly.JellyFacet.handleIndexRequest(JellyFacet.java:127)
	at org.kohsuke.stapler.Stapler.tryInvoke(Stapler.java:735)
	at org.kohsuke.stapler.Stapler.invoke(Stapler.java:876)
	at org.kohsuke.stapler.MetaClass$10.dispatch(MetaClass.java:362)
	at org.kohsuke.stapler.Stapler.tryInvoke(Stapler.java:746)
	at org.kohsuke.stapler.Stapler.invoke(Stapler.java:876)
	at org.kohsuke.stapler.MetaClass$5.doDispatch(MetaClass.java:236)
	at org.kohsuke.stapler.NameBasedDispatcher.dispatch(NameBasedDispatcher.java:58)
	at org.kohsuke.stapler.Stapler.tryInvoke(Stapler.java:746)
	at org.kohsuke.stapler.Stapler.invoke(Stapler.java:876)
	at org.kohsuke.stapler.Stapler.invoke(Stapler.java:649)
	at org.kohsuke.stapler.Stapler.service(Stapler.java:238)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:790)
	at org.eclipse.jetty.servlet.ServletHolder.handle(ServletHolder.java:812)
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1669)
	at hudson.util.PluginServletFilter$1.doFilter(PluginServletFilter.java:135)
	at hudson.util.PluginServletFilter.doFilter(PluginServletFilter.java:126)
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1652)
	at hudson.security.csrf.CrumbFilter.doFilter(CrumbFilter.java:86)
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1652)
	at hudson.security.ChainedServletFilter$1.doFilter(ChainedServletFilter.java:84)
	at hudson.security.UnwrapSecurityExceptionFilter.doFilter(UnwrapSecurityExceptionFilter.java:51)
	at hudson.security.ChainedServletFilter$1.doFilter(ChainedServletFilter.java:87)
	at jenkins.security.ExceptionTranslationFilter.doFilter(ExceptionTranslationFilter.java:117)
	at hudson.security.ChainedServletFilter$1.doFilter(ChainedServletFilter.java:87)
	at org.acegisecurity.providers.anonymous.AnonymousProcessingFilter.doFilter(AnonymousProcessingFilter.java:125)
	at hudson.security.ChainedServletFilter$1.doFilter(ChainedServletFilter.java:87)
	at org.acegisecurity.ui.rememberme.RememberMeProcessingFilter.doFilter(RememberMeProcessingFilter.java:142)
	at hudson.security.ChainedServletFilter$1.doFilter(ChainedServletFilter.java:87)
	at org.acegisecurity.ui.AbstractProcessingFilter.doFilter(AbstractProcessingFilter.java:271)
	at hudson.security.ChainedServletFilter$1.doFilter(ChainedServletFilter.java:87)
	at jenkins.security.BasicHeaderProcessor.doFilter(BasicHeaderProcessor.java:93)
	at hudson.security.ChainedServletFilter$1.doFilter(ChainedServletFilter.java:87)
	at org.acegisecurity.context.HttpSessionContextIntegrationFilter.doFilter(HttpSessionContextIntegrationFilter.java:249)
	at hudson.security.HttpSessionContextIntegrationFilter2.doFilter(HttpSessionContextIntegrationFilter2.java:67)
	at hudson.security.ChainedServletFilter$1.doFilter(ChainedServletFilter.java:87)
	at hudson.security.ChainedServletFilter.doFilter(ChainedServletFilter.java:76)
	at hudson.security.HudsonFilter.doFilter(HudsonFilter.java:171)
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1652)
	at org.kohsuke.stapler.compression.CompressionFilter.doFilter(CompressionFilter.java:49)
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1652)
	at hudson.util.CharacterEncodingFilter.doFilter(CharacterEncodingFilter.java:82)
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1652)
	at org.kohsuke.stapler.DiagnosticThreadNameFilter.doFilter(DiagnosticThreadNameFilter.java:30)
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1652)
	at org.eclipse.jetty.servlet.ServletHandler.doHandle(ServletHandler.java:585)
	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:143)
	at org.eclipse.jetty.security.SecurityHandler.handle(SecurityHandler.java:553)
	at org.eclipse.jetty.server.session.SessionHandler.doHandle(SessionHandler.java:223)
	at org.eclipse.jetty.server.handler.ContextHandler.doHandle(ContextHandler.java:1127)
	at org.eclipse.jetty.servlet.ServletHandler.doScope(ServletHandler.java:515)
	at org.eclipse.jetty.server.session.SessionHandler.doScope(SessionHandler.java:185)
	at org.eclipse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:1061)
	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:141)
	at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:97)
	at org.eclipse.jetty.server.Server.handle(Server.java:499)
	at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:311)
	at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:257)
	at org.eclipse.jetty.io.AbstractConnection$2.run(AbstractConnection.java:544)
	at winstone.BoundedExecutorService$1.run(BoundedExecutorService.java:77)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:748)
Caused by: org.apache.commons.jelly.JellyTagException: jar:file:/var/lib/jenkins/plugins/build-time-blame/WEB-INF/lib/build-time-blame.jar!/org/jenkins/ci/plugins/buildtimeblame/action/BlameAction/index.jelly:11:70: <st:include> No page found 'body.jelly' for class org.jenkins.ci.plugins.buildtimeblame.action.BlameAction
	at org.kohsuke.stapler.jelly.IncludeTag.doTag(IncludeTag.java:124)
	at org.apache.commons.jelly.impl.TagScript.run(TagScript.java:269)
	at org.apache.commons.jelly.impl.ScriptBlock.run(ScriptBlock.java:95)
	at org.apache.commons.jelly.tags.core.CoreTagLibrary$1.run(CoreTagLibrary.java:98)
	at org.apache.commons.jelly.impl.ScriptBlock.run(ScriptBlock.java:95)
	at org.kohsuke.stapler.jelly.ReallyStaticTagLibrary$1.run(ReallyStaticTagLibrary.java:99)
	at org.kohsuke.stapler.jelly.CallTagLibScript$1.run(CallTagLibScript.java:99)
	at org.apache.commons.jelly.tags.define.InvokeBodyTag.doTag(InvokeBodyTag.java:91)
	at org.apache.commons.jelly.impl.TagScript.run(TagScript.java:269)
	at org.apache.commons.jelly.impl.ScriptBlock.run(ScriptBlock.java:95)
	at org.apache.commons.jelly.tags.core.CoreTagLibrary$1.run(CoreTagLibrary.java:98)
	at org.apache.commons.jelly.impl.ScriptBlock.run(ScriptBlock.java:95)
	at org.apache.commons.jelly.tags.core.CoreTagLibrary$2.run(CoreTagLibrary.java:105)
	at org.kohsuke.stapler.jelly.CallTagLibScript.run(CallTagLibScript.java:120)
	at org.apache.commons.jelly.impl.ScriptBlock.run(ScriptBlock.java:95)
	at org.kohsuke.stapler.jelly.CallTagLibScript$1.run(CallTagLibScript.java:99)
	at org.apache.commons.jelly.tags.define.InvokeBodyTag.doTag(InvokeBodyTag.java:91)
	at org.apache.commons.jelly.impl.TagScript.run(TagScript.java:269)
	at org.apache.commons.jelly.impl.ScriptBlock.run(ScriptBlock.java:95)
	at org.kohsuke.stapler.jelly.ReallyStaticTagLibrary$1.run(ReallyStaticTagLibrary.java:99)
	at org.apache.commons.jelly.impl.ScriptBlock.run(ScriptBlock.java:95)
	at org.kohsuke.stapler.jelly.ReallyStaticTagLibrary$1.run(ReallyStaticTagLibrary.java:99)
	at org.apache.commons.jelly.impl.ScriptBlock.run(ScriptBlock.java:95)
	at org.kohsuke.stapler.jelly.ReallyStaticTagLibrary$1.run(ReallyStaticTagLibrary.java:99)
	at org.apache.commons.jelly.impl.ScriptBlock.run(ScriptBlock.java:95)
	at org.kohsuke.stapler.jelly.ReallyStaticTagLibrary$1.run(ReallyStaticTagLibrary.java:99)
	at org.apache.commons.jelly.impl.ScriptBlock.run(ScriptBlock.java:95)
	at org.apache.commons.jelly.tags.core.CoreTagLibrary$2.run(CoreTagLibrary.java:105)
	at org.kohsuke.stapler.jelly.CallTagLibScript.run(CallTagLibScript.java:120)
	at org.apache.commons.jelly.tags.core.CoreTagLibrary$2.run(CoreTagLibrary.java:105)
	at org.kohsuke.stapler.jelly.JellyViewScript.run(JellyViewScript.java:95)
	at org.kohsuke.stapler.jelly.DefaultScriptInvoker.invokeScript(DefaultScriptInvoker.java:63)
	at org.kohsuke.stapler.jelly.DefaultScriptInvoker.invokeScript(DefaultScriptInvoker.java:53)
	at org.kohsuke.stapler.jelly.JellyClassTearOff.serveIndexJelly(JellyClassTearOff.java:112)
	... 63 more
```

```
	org.jenkins-ci.main:jenkins-war:2.46.3
```

--- a/src/main/groovy/org/jenkins/ci/plugins/buildtimeblame/action/BlameAction.groovy
+++ b/src/main/groovy/org/jenkins/ci/plugins/buildtimeblame/action/BlameAction.groovy
@@ -110,7 +110,7 @@ class BlameAction implements Action {
         }.collect { Run run ->
             try {
                 return logParser.getBuildResult(run)
-            } catch (LogParser.TimestampMissingException ignored) {
+            } catch (RuntimeException ignored) {
                 buildsWithoutTimestamps.add(run)
                 return null
             } finally {

--- a/src/main/groovy/org/jenkins/ci/plugins/buildtimeblame/analysis/LogParser.groovy
+++ b/src/main/groovy/org/jenkins/ci/plugins/buildtimeblame/analysis/LogParser.groovy
@@ -2,6 +2,7 @@
 package org.jenkins.ci.plugins.buildtimeblame.analysis
 
 import com.google.common.base.Optional
+import com.thoughtworks.xstream.mapper.Mapper.Null
 import hudson.model.Run
 import hudson.plugins.timestamper.Timestamp
 import hudson.plugins.timestamper.io.TimestampsReader
@@ -47,7 +48,7 @@ class LogParser {
 
     private void processMatches(Run run, Closure onMatch) {
         def numberOfMissingTimestamps = 0
-        Timestamp previousTimestamp
+        Timestamp mostRecentTimestamp
 
         def timestampsReader = new TimestampsReader(run)
         def steps = relevantSteps.collect()
@@ -57,7 +58,7 @@ class LogParser {
             def step = getMatchingRegex(line, steps)
 
             if (nextTimestamp.isPresent()) {
-                previousTimestamp = nextTimestamp.get()
+                mostRecentTimestamp = nextTimestamp.get()
             } else {
                 numberOfMissingTimestamps++
             }
@@ -66,9 +67,9 @@ class LogParser {
                 if (nextTimestamp.isPresent()) {
                     def timestamp = nextTimestamp.get()
                     onMatch(step.get().label, line, timestamp)
-                    previousTimestamp = nextTimestamp.get()
-                } else if (numberOfMissingTimestamps <= maximumMissingTimestamps) {
-                    onMatch(step.get().label, line, previousTimestamp)
+                    mostRecentTimestamp = nextTimestamp.get()
+                } else if (mostRecentTimestamp && numberOfMissingTimestamps <= maximumMissingTimestamps) {
+                    onMatch(step.get().label, line, mostRecentTimestamp)
                 } else {
                     throw new TimestampMissingException()
                 }

--- a/src/main/resources/org/jenkins/ci/plugins/buildtimeblame/action/BlameAction/body.jelly
+++ b/src/main/resources/org/jenkins/ci/plugins/buildtimeblame/action/BlameAction/body.jelly
@@ -1,0 +1,4 @@
+<?jelly escape-by-default='true'?>
+<j:jelly xmlns:j="jelly:core" xmlns:st="jelly:stapler" xmlns:d="jelly:define" xmlns:l="/lib/layout"
+         xmlns:t="/lib/hudson" xmlns:f="/lib/form">
+</j:jelly>

--- a/src/test/groovy/org/jenkins/ci/plugins/buildtimeblame/action/BlameActionTest.groovy
+++ b/src/test/groovy/org/jenkins/ci/plugins/buildtimeblame/action/BlameActionTest.groovy
@@ -156,7 +156,7 @@ class BlameActionTest extends Specification {
         1 * project.getBuilds() >> RunList.fromRuns([build1, build2])
         1 * new LogParser(blameAction.relevantSteps) >> logParser
         1 * logParser.getBuildResult(build1) >> build1Results
-        1 * logParser.getBuildResult(build2) >> { throw new LogParser.TimestampMissingException() }
+        1 * logParser.getBuildResult(build2) >> { throw new RuntimeException() }
         report == new BlameReport([build1Results])
         blameAction.buildsWithoutTimestamps == [build2]
         blameAction.lastProcessedBuild == lastBuildNumber

--- a/src/test/groovy/org/jenkins/ci/plugins/buildtimeblame/analysis/LogParserTest.groovy
+++ b/src/test/groovy/org/jenkins/ci/plugins/buildtimeblame/analysis/LogParserTest.groovy
@@ -157,6 +157,22 @@ class LogParserTest extends Specification {
         thrown(LogParser.TimestampMissingException)
     }
 
+    def 'should throw error if there are no timestamps'() {
+        given:
+        TimestampsReader timestampReader = GroovyMock(TimestampsReader, global: true)
+        def build = Mock(Run)
+        setupMockLog(build, 'line1', 'line2', 'line3', 'line4')
+        def logParser = new LogParser([new RelevantStep(~/line1/, '', false)])
+
+        when:
+        logParser.getBuildResult(build)
+
+        then:
+        _ * new TimestampsReader(build) >> timestampReader
+        1 * timestampReader.read() >> Optional.absent()
+        thrown(LogParser.TimestampMissingException)
+    }
+
     def 'should ignore missing timestamps if no match is found after them'() {
         given:
         TimestampsReader timestampReader = GroovyMock(TimestampsReader, global: true)
@@ -210,7 +226,7 @@ class LogParserTest extends Specification {
         results.build == build
     }
 
-    def 'should have expected default number of missing timestamps'() {
+    def 'should have expected default number of ignored missing timestamps'() {
         given:
         def logParser = new LogParser([])
 

